Docs: Add testing guide (#263)
aragon/aragon-court: docs/8-testing-guide/readme.md

# 8. Testing guide

This guide aims to cover all the things you should know in order to try Aragon Court or integrate your application with it. 

## 8.1. Testing instances

There are a few testing instances already deployed for Aragon Court. 

> All these are mimicking the instance living on mainnet with the exception of some dispute periods durations to provide a better testing experience.

Reviewer: ```suggestion
All of these are mimicking the mainnet instance with some exceptions of term durations to provide a better testing experience.
```

> Additionally, all these instances are using their own deployed version of the following ERC20 tokens:

Reviewer: ```suggestion
Additionally, all the instances are using their own deployed version of the following ERC20 tokens:
```

- ANJ, the native token of Aragon Court. You will need some fake ANJ to stake as a juror to be selected to resolve disputes.  
- DAI, used for the Aragon Court fees. You will need some fake DAI to pay jurors and pay the Court subscription fees.

> Of course, there is an ERC20 faucet deployed for all these instances that you can use to claim some fake ANJ or DAI to start testing.
> Please keep reading to understand better how to use the faucet along with the Aragon Court testing instance. 

Reviewer: ```suggestion
Of course, there is an ERC20 faucet deployed for all these instances that you can use to claim some fake ANJ or DAI to start testing. More information is outlined below on using these faucets.
```

- ERC20 faucet: [`0x3b86Fd8C30445Ddcbed79CE7eB052fe935D34Fd2`](http://rinkeby.etherscan.io/address/0x3b86Fd8C30445Ddcbed79CE7eB052fe935D34Fd2)

### 8.1.3. Ropsten

- Network: Ropsten
- Court term: 8 hours (same as mainnet)
- Subscription period: 720 court terms
- Dashboard: https://court-ropsten.aragon.org/
- Address: [`0x3b26bc496aebaed5b3e0e81cde6b582cde71396e`](http://ropsten.etherscan.io/address/0x3b26bc496aebaed5b3e0e81cde6b582cde71396e)
- Fake ANJ: [`0xc863e1ccc047beff17022f4229dbe6321a6bce65`](http://ropsten.etherscan.io/address/0xc863e1ccc047beff17022f4229dbe6321a6bce65)
- Fake DAI: [`0x4e1f48db14d7e1ada090c42ffe15ff3024eec8bf`](http://ropsten.etherscan.io/address/0x4e1f48db14d7e1ada090c42ffe15ff3024eec8bf)
- ERC20 faucet: [`0x83c1ECDC6fAAb783d9e3ac2C714C0eEce3349638`](http://ropsten.etherscan.io/address/0x83c1ECDC6fAAb783d9e3ac2C714C0eEce3349638)

### 8.1.4. Local

To deploy a local instance of Aragon Court you will need to clone the deployment scripts first:
 
```bash
git clone https://github.com/aragon/aragon-network-deploy/
cd aragon-network-deploy
npm i

> npm run deploy:court:rpc 

Reviewer 2: Got this error at this step, not sure what to do next.

```
npm run deploy:court:rpc

> @aragon/court-deploy@1.2.0 deploy:court:rpc /home/light/aragon-network-deploy
> rm -f data/output/{minime,court}.rpc.json && npm run deploy:dai -- -n rpc && npm run deploy:anj -- -n rpc && npm run deploy:court -- -n rpc

> @aragon/court-deploy@1.2.0 deploy:dai /home/light/aragon-network-deploy
> node ./bin/index.js deploy-minime -i ./test/data/input/dai.js --verbose "-n" "rpc"

Error: Cannot find module '@truffle/debug-utils'
    at Function.Module._resolveFilename (internal/modules/cjs/loader.js:582:15)
    at Function.Module._load (internal/modules/cjs/loader.js:508:25)
    at Module.require (internal/modules/cjs/loader.js:637:17)
    at require (internal/modules/cjs/helpers.js:22:18)
    at Object.<anonymous> (/home/light/aragon-network-deploy/node_modules/@truffle/contract/lib/promievent.js:2:20)
    at Module._compile (internal/modules/cjs/loader.js:701:30)
    at Object.Module._extensions..js (internal/modules/cjs/loader.js:712:10)
    at Module.load (internal/modules/cjs/loader.js:600:32)
    at tryModuleLoad (internal/modules/cjs/loader.js:539:12)
    at Function.Module._load (internal/modules/cjs/loader.js:531:3)
    at Module.require (internal/modules/cjs/loader.js:637:17)
    at require (internal/modules/cjs/helpers.js:22:18)
    at Object.<anonymous> (/home/light/aragon-network-deploy/node_modules/@truffle/contract/lib/execute.js:2:20)
    at Module._compile (internal/modules/cjs/loader.js:701:30)
    at Object.Module._extensions..js (internal/modules/cjs/loader.js:712:10)
    at Module.load (internal/modules/cjs/loader.js:600:32)
npm ERR! code ELIFECYCLE
npm ERR! errno 1
npm ERR! @aragon/court-deploy@1.2.0 deploy:dai: `node ./bin/index.js deploy-minime -i ./test/data/input/dai.js --verbose "-n" "rpc"`
npm ERR! Exit status 1
npm ERR! 
npm ERR! Failed at the @aragon/court-deploy@1.2.0 deploy:dai script.
npm ERR! This is probably not a problem with npm. There is likely additional logging output above.

npm ERR! A complete log of this run can be found in:
npm ERR!     /home/light/.npm/_logs/2020-05-01T21_10_33_463Z-debug.log
npm ERR! code ELIFECYCLE
npm ERR! errno 1
npm ERR! @aragon/court-deploy@1.2.0 deploy:court:rpc: `rm -f data/output/{minime,court}.rpc.json && npm run deploy:dai -- -n rpc && npm run deploy:anj -- -n rpc && npm run deploy:court -- -n rpc`
npm ERR! Exit status 1
npm ERR! 
npm ERR! Failed at the @aragon/court-deploy@1.2.0 deploy:court:rpc script.
npm ERR! This is probably not a problem with npm. There is likely additional logging output above.
```

Author: Unfortunately, Truffle lib is broken there is an issue open for this

Author: I committed a patch for this pinning a version of truffle contract that actually works, you can try again now

Reviewer 2: @facuspagnuolo deleted the old aragon-network-deploy directory and re-downloaded, getting a new error now:

```
npm run deploy:court:rpc

> @aragon/court-deploy@1.2.0 deploy:court:rpc /home/light/aragon-network-deploy
> rm -f data/output/{minime,court}.rpc.json && npm run deploy:dai -- -n rpc && npm run deploy:anj -- -n rpc && npm run deploy:court -- -n rpc

> @aragon/court-deploy@1.2.0 deploy:dai /home/light/aragon-network-deploy
> node ./bin/index.js deploy-minime -i ./test/data/input/dai.js --verbose "-n" "rpc"

[Error Handler]                 🚨  Process finished with error:
{ Error: 
Could not connect to your Ethereum client with the following parameters:
    - host       > localhost
    - port       > 8545
    - network_id > 15
Please check that your Ethereum client:
    - is running
    - is accepting RPC connections (i.e., "--rpc" option is used in geth)
    - is accessible over the network
    - is properly configured in your Truffle configuration file (truffle-config.js)

    at ProviderError.ExtendableError (/home/light/aragon-network-deploy/node_modules/@truffle/error/index.js:10:16)
    at new ProviderError (/home/light/aragon-network-deploy/node_modules/@truffle/provider/error.js:12:24)
    at /home/light/aragon-network-deploy/node_modules/@truffle/provider/wrapper.js:75:17
    at /home/light/aragon-network-deploy/node_modules/@truffle/provider/wrapper.js:107:24
    at XMLHttpRequest.request.onreadystatechange (/home/light/aragon-network-deploy/node_modules/web3-providers-http/src/index.js:96:13)
    at XMLHttpRequestEventTarget.dispatchEvent (/home/light/aragon-network-deploy/node_modules/xhr2-cookies/xml-http-request-event-target.ts:44:13)
    at XMLHttpRequest._setReadyState (/home/light/aragon-network-deploy/node_modules/xhr2-cookies/xml-http-request.ts:219:8)
    at XMLHttpRequest._onHttpRequestError (/home/light/aragon-network-deploy/node_modules/xhr2-cookies/xml-http-request.ts:379:8)
    at ClientRequest.<anonymous> (/home/light/aragon-network-deploy/node_modules/xhr2-cookies/xml-http-request.ts:266:37)
    at ClientRequest.emit (events.js:189:13)
    at Socket.socketErrorListener (_http_client.js:392:9)
    at Socket.emit (events.js:189:13)
    at emitErrorNT (internal/streams/destroy.js:82:8)
    at emitErrorAndCloseNT (internal/streams/destroy.js:50:3)
    at process._tickCallback (internal/process/next_tick.js:63:19)
  message:
   '\nCould not connect to your Ethereum client with the following parameters:\n    - host       > localhost\n    - port       > 8545\n    - network_id > 15\nPlease check that your Ethereum client:\n    - is running\n    - is accepting RPC connections (i.e., "--rpc" option is used in geth)\n    - is accessible over the network\n    - is properly configured in your Truffle configuration file (truffle-config.js)\n',
  stack:
   'Error: \nCould not connect to your Ethereum client with the following parameters:\n    - host       > localhost\n    - port       > 8545\n    - network_id > 15\nPlease check that your Ethereum client:\n    - is running\n    - is accepting RPC connections (i.e., "--rpc" option is used in geth)\n    - is accessible over the network\n    - is properly configured in your Truffle configuration file (truffle-config.js)\n\n    at ProviderError.ExtendableError (/home/light/aragon-network-deploy/node_modules/@truffle/error/index.js:10:16)\n    at new ProviderError (/home/light/aragon-network-deploy/node_modules/@truffle/provider/error.js:12:24)\n    at /home/light/aragon-network-deploy/node_modules/@truffle/provider/wrapper.js:75:17\n    at /home/light/aragon-network-deploy/node_modules/@truffle/provider/wrapper.js:107:24\n    at XMLHttpRequest.request.onreadystatechange (/home/light/aragon-network-deploy/node_modules/web3-providers-http/src/index.js:96:13)\n    at XMLHttpRequestEventTarget.dispatchEvent (/home/light/aragon-network-deploy/node_modules/xhr2-cookies/xml-http-request-event-target.ts:44:13)\n    at XMLHttpRequest._setReadyState (/home/light/aragon-network-deploy/node_modules/xhr2-cookies/xml-http-request.ts:219:8)\n    at XMLHttpRequest._onHttpRequestError (/home/light/aragon-network-deploy/node_modules/xhr2-cookies/xml-http-request.ts:379:8)\n    at ClientRequest.<anonymous> (/home/light/aragon-network-deploy/node_modules/xhr2-cookies/xml-http-request.ts:266:37)\n    at ClientRequest.emit (events.js:189:13)\n    at Socket.socketErrorListener (_http_client.js:392:9)\n    at Socket.emit (events.js:189:13)\n    at emitErrorNT (internal/streams/destroy.js:82:8)\n    at emitErrorAndCloseNT (internal/streams/destroy.js:50:3)\n    at process._tickCallback (internal/process/next_tick.js:63:19)',
  name: 'ProviderError' }

npm ERR! code ELIFECYCLE
npm ERR! errno 1
npm ERR! @aragon/court-deploy@1.2.0 deploy:dai: `node ./bin/index.js deploy-minime -i ./test/data/input/dai.js --verbose "-n" "rpc"`
npm ERR! Exit status 1
npm ERR! 
npm ERR! Failed at the @aragon/court-deploy@1.2.0 deploy:dai script.
npm ERR! This is probably not a problem with npm. There is likely additional logging output above.

npm ERR! A complete log of this run can be found in:
npm ERR!     /home/light/.npm/_logs/2020-05-04T21_16_36_103Z-debug.log
npm ERR! code ELIFECYCLE
npm ERR! errno 1
npm ERR! @aragon/court-deploy@1.2.0 deploy:court:rpc: `rm -f data/output/{minime,court}.rpc.json && npm run deploy:dai -- -n rpc && npm run deploy:anj -- -n rpc && npm run deploy:court -- -n rpc`
npm ERR! Exit status 1
npm ERR! 
npm ERR! Failed at the @aragon/court-deploy@1.2.0 deploy:court:rpc script.
npm ERR! This is probably not a problem with npm. There is likely additional logging output above.
```

Author: Hey, that's because we need a local ganache running to deploy it. I added one more step to make that clear

npm run deploy:court:rpc 
```

Once you have done that, you can deploy a local instance by running the following command:

```bash
npm run deploy:court:rpc
```

> This command will output the addresses of all the deployed modules of Aragon Court including the main entry point.

Reviewer: ```suggestion
This command will output the addresses of all the deployed modules of Aragon Court including the main entry point (the `AragonCourt` contract).
```

> Additionally, it should deploy a fake version of the ANJ and DAI tokens to be used for testing purposes as explained above.

Reviewer: ```suggestion
Additionally, it should deploy a fake version of the ANJ and DAI tokens usable for testing purposes as explained above.
```

## 8.2. Claiming fake tokens from the ERC20 faucets

You can claim ANJ or DAI fake tokens from the ERC20 faucets.
You can do this directly through Etherscan, simply click in any of the faucet links shared above in section 8.1.

> Once there, you just need to enable your Web3 account and call the `withdraw()` function providing the desired token address and amount:

Reviewer 2: There is no `withdraw()` function on the Rinkeby ANJ contract:

https://rinkeby.etherscan.io/address/0x975ef6b5fde81c24c4ec605091f2e945872b6036#writeContract

Reviewer 2: ![Screenshot_2020-05-01 Contract Address 0x975ef6b5fde81c24c4ec605091f2e945872b6036 Etherscan](https://user-images.githubusercontent.com/9424721/80842885-bdae5a00-8bd0-11ea-8573-187276014c61.png)

Reviewer 2: Ok I figured out I was looking at the wrong contract.

> Once there, you just need to enable your Web3 account and call the `withdraw()` function providing the desired token address and amount:

Reviewer 2: > providing the desired token address and amount

if I want 1000 tokens do I just enter 1000 or do I have to add 18 zeroes to account for the decimal places?

Reviewer 2: I entered `10000` since that was the limit and the transaction failed with this error message:

> Fail with error 'FAUCET_NOT_ENOUGH_SUPPLY'

https://rinkeby.etherscan.io/tx/0x1b51a1421522214e1e81358eff6e5fd005efd19d83cc90445fc91d0ea76b0df5

Reviewer 2: Could be because the faucet contract doesn't hold any ANJ 🤔

![anjfail](https://user-images.githubusercontent.com/9424721/80844773-87270e00-8bd5-11ea-88fc-765554f83841.png)

Reviewer 2: Without ANJ to test with I'm not sure how to proceed with testing this guide. Stopping here and will come back once I have some test ANJ.

Author: It was because the faucet didn't have enough ANJ, I funded it now

Reviewer 2: I think I answered my own question above: 

> if I want 1000 tokens do I just enter 1000 or do I have to add 18 zeroes to account for the decimal places?

I had to enter 9000 + 18 zeroes to get 9000 ANJ.

> Once in there, you just need to enable your Web3 account and call the `withdraw` function providing the desired token address and amount:

Reviewer: ```suggestion
Once there, you just need to enable your Web3 account and call the `withdraw()` function providing the desired token address and amount:
```

![faucet](./faucet.png)

> Bear in mind there is a quota set for these faucets, all of them will allow you to withdraw up to 10,000 fake-DAI or 10,000 fake-ANJ every 7 days.

Reviewer: ```suggestion
Bear in mind there is a quota set for these faucets; they will only allow you to withdraw up to 10,000 fake-DAI or 10,000 fake-ANJ every 7 days.
```

## 8.3. Installing the Aragon Court dev CLI tool

To interact with the deployed versions of Aragon Court, we built a node-based [CLI tool](https://github.com/aragonone/court-backend/tree/development/packages/cli) that you can use.
Currently, there is no published version of it. But you can clone the GitHub repo and run it locally.
To continue with the testing guide you will need to use it. First, make sure you clone it and install its dependencies as follows:
```
git clone https://github.com/aragonone/court-backend/
cd court-backend
npm i
npx lerna bootstrap
cd packages/cli
```

This CLI tool is built on top of Truffle using a custom [config file](https://www.npmjs.com/package/@aragon/truffle-config-v5) provided by Aragon.

> Please check it out to make sure how to set up your private keys for testing.

Reviewer: ```suggestion
Please review that package's documentation to understand how to set up your private keys for testing.
```

> Let's continue with with the Aragon Court testing guide and see how we can use the CLI tool.

Reviewer 2: ```suggestion
Let's continue with the Aragon Court testing guide and see how we can use the CLI tool.
```

Let's continue with with the Aragon Court testing guide and see how we can use the CLI tool.

## 8.4. Becoming a juror

To become a juror you simply need to activate some ANJ tokens into Aragon Court.
First make sure to have claimed some fake ANJ tokens from the faucet corresponding to the Aragon Court instance you're willing to try. 
Then, you can activate tokens into Aragon Court using the `activate` command of the CLI tool as follows:

```bash
node ./bin/index.js activate --jurors [JUROR] --amount [AMOUNT] --from [FROM] --network [NETWORK] 
```

Where:

> - `[JUROR]` is the address of the jurors you will activate the tokens for
> - `[AMOUNT]` is the amount of fake ANJ tokens you will activate for the target juror
> - `[FROM]` is the address paying for the fake ANJ tokens, this must be the address you used to claim the tokens from the faucet
> - `[NETWORK]` is the name of the Aragon Court instance you are willing to use: `usability`, `rinkeby`, or `ropsten` 

Reviewer: ```suggestion
- `[JUROR]`: address of the juror you will activate the tokens for
- `[AMOUNT]`: amount of fake ANJ tokens you will activate for the specified juror
- `[FROM]`: address paying for the fake ANJ tokens; this must be the address you used to claim the tokens from the faucet
- `[NETWORK]`: name of the Aragon Court instance you are willing to use: `usability`, `rinkeby`, or `ropsten` 
```

> In case the sender address is the jurors, you can simplify this a bit as follows:

Reviewer: ```suggestion
In case the sender address is the juror to be activated, you can simplify to the following:
```

```bash
node ./bin/index.js activate -a [AMOUNT] -n [NETWORK]
```

Note that you can also use the flag `--verbose` to have more details about the transactions being sent to the network.

You can check your current stake as a juror in the dashboards linked above in section 8.1.

## 8.5. Creating a dispute

> As you may know, based on how Aragon Court works, disputes must be created from a particular type of smart contracts that support being ruled by the court itself.

Reviewer: ```suggestion
As you may know, disputes can only be submitted to Aragon Court through smart contracts that implement a specific interface to support being ruled by the court itself.
```

This is specified by the [`IArbitrable`](../../contracts/arbitration/IArbitrable.sol) interface.

Thus, the first thing we should do is to deploy an Arbitrable contract. You can do this from the CLI running the following command:

```bash
node ./bin/index.js arbitrable -n [NETWORK]
```

This command will output the address of your new Arbitrable contract.

> The next we should do is to subscribe your Arbitrable instance to Aragon Court paying the subscription fees.

Reviewer: ```suggestion
The next step is to subscribe your Arbitrable instance to Aragon Court by paying the subscription fees.
```

For the testing instances, each subscription period costs 7,500 fake-DAI, so make sure you claim some fake DAI from the ERC20 faucet as described in section 8.2.

Once you have done that you can subscribe your Arbitrable instance running the following command:

```bash
node ./bin/index.js subscribe -a [ARBITRABLE] -n [NETWORK]
``` 

Where `[ARBITRABLE]` is the address of your Arbitrable instance.

> Now, we are almost ready to create a dispute, but first we will need to have some fake DAI in the Arbitrable instance to pay for the dispute fees. 

Reviewer: ```suggestion
Now, we are almost ready to create a dispute. The last step is to send some fake DAI to the Arbitrable instance so that it can pay for the court's dispute fees.
```

These are different from the subscription fees. The dispute fees are to pay the jurors for each dispute to be resolved.
For the testing instances, each dispute costs 30.87 fake-DAI.
Thus, you will need to make a transfer from your account to your Arbitrable instance.
To do that you can use the Etherscan interface for the DAI instance linked in section 8.1.

Finally, we are ready to create your dispute running the following command:

```bash
node ./bin/index.js dispute \
  -a [ARBITRABLE] \
  -m [METADATA] \
  -e [EVIDENCE_1] [EVIDENCE_2] ... [EVIDENCE_N] \
  -s [SUBMITTER_1] [SUBMITTER_1] ... [SUBMITTER_N] \
  -c \
  -n [NETWORK] \
```

Where: 

> - `[ARBITRABLE]` is the address of your Arbitrable instance
> - `[METADATA]` is the metadata to be linked for your dispute (continue reading to have a better understanding of how to build a proper dispute metadata)
> - `[EVIDENCE_N]` are IPFS links in the form of `ipfs:[HASH]` to human-redeable evidence (continue reading to find some examples about evidences) 
> - `[SUBMITTER_N]` are the addresses submitting each evidence piece of information, this list should match the evidence list length 
> - The `-c` flag is optional to declare that the evidence submission period is close. Otherwise you will need to close it manually afterwards. 
> - `[NETWORK]` is the name of the Aragon Court instance you are willing to use: `usability`, `rinkeby`, or `ropsten` 

Reviewer: ```suggestion
- `[ARBITRABLE]`: address of your Arbitrable instance
- `[METADATA]`: metadata to be linked for your dispute (continue reading to have a better understanding of how to build a proper dispute metadata)
- `[EVIDENCE_N]`: links in the form of `ipfs:[CID]` to human-readable evidence hosted on IPFS (continue reading for example evidence links) 
- `[SUBMITTER_N]`: addresses submitting each piece of evidence; this list should match the evidence list length 
- `-c` flag: optional to declare that the evidence submission period should be immediately closed. Otherwise you will need to manually close it afterwards. 
- `[NETWORK]`: name of the Aragon Court instance you are willing to use: `usability`, `rinkeby`, or `ropsten`
```

This command will output the ID of the dispute you've just created.

A few things to bear in mind is that, even though the `[METADATA]` and `[EVIDENCE_N]` arguments could be any arbitrary information, in order to use the Court Dashboard to rule disputes, these should follow a specific structure.

### 8.5.1. Metadata

> The Court Dashboard expects the metadata to be built as a specific JSON formatted data:

Reviewer: We may instead want to [link to the court dashboard repo](https://github.com/aragon/court-dashboard/blob/development/docs/metadata.md) (and update it if need be), since this is more of a dashboard-level responsibility.

```json
{ 
    "description": "Your dispute description",  
    "metadata": "[IPFS_HASH]/metadata.json" 
}
```

Where `[IPFS_HASH]` is the hash of a directory holding a `metadata.json` file with the rest of the metadata.
You can checkout the [`metadata.json`](https://ipfs.io/ipfs/QmYt33BkuHMLe4dSTfLan7QXxQVTGRyhVLt5sujrZkhd1w/metadata.json) file used for the [dispute #1](https://court-usability.aragon.org/disputes/1) of Aragon Court as an example.

### 8.5.2. Evidence

For the Court Dashboard, the evidence is simpler than the metadata, it only needs to be a human-readable content uploaded to IPFS.
Then it should be submitted as `ipfs:[HASH]`. For example, [`ipfs:QmYGNe8jhTEwdDfixtDnPpzjQpXhX2nMj3xMK3swy69naP`](https://ipfs.io/ipfs/QmYGNe8jhTEwdDfixtDnPpzjQpXhX2nMj3xMK3swy69naP) is the evidence submitted for the [dispute #1](https://court-usability.aragon.org/disputes/1) of Aragon Court.

## 8.6. Ruling a dispute

In case you created the dispute in way that is compatible with the Court Dashboard as explained in sections 8.5.1 and 8.5.2, you can use any of the Court Dashboard instances linked in section 8.1 to rule your disputes.

> Otherwise, you still can use the Court Dashboard for it but you probably won't be able to read your metadata or evidence properly.

Reviewer: You can use any of the Court Dashboard instances linked in section 8.1 to interact with your created disputes (note that in some environments, it may be difficult to ensure that your account is drafted—and therefore can be difficult to come to a ruling you want—due to the randomness nature of the court). If you did not format or ensure your dispute metadata was available as explained in sections 8.5.1 and 8.5.2, the dispute will most likely not display the intended information to jurors.

> Alternatively, you can use the rest of the CLI tool [commands](https://github.com/aragonone/court-backend/tree/master/packages/cli/#commands) to start ruling your dispute:

Reviewer: ```suggestion
Alternatively, you can use the rest of the CLI tool [commands](https://github.com/aragonone/court-backend/tree/master/packages/cli/#commands) to begin ruling your dispute:
```

Currently, there is no published version of it. But you can clone the GitHub repo and run it locally.
To continue with the testing guide you will need to use it. First, make sure you clone it and install its dependencies as follows:
```
git clone https://github.com/aragonone/court-backend/
cd court-backend
npm i
npx lerna bootstrap
cd packages/cli
```

This CLI tool is built on top of Truffle using a custom [config file](https://www.npmjs.com/package/@aragon/truffle-config-v5) provided by Aragon.
Please review that package's documentation to understand how to set up your private keys for testing.

Let's continue with the Aragon Court testing guide and see how we can use the CLI tool.

## 8.4. Becoming a juror

To become a juror you simply need to activate some ANJ tokens into Aragon Court.
First make sure to have claimed some fake ANJ tokens from the faucet corresponding to the Aragon Court instance you're willing to try. 
Then, you can activate tokens into Aragon Court using the `activate` command of the CLI tool as follows:

```bash

> node ./bin/index.js activate --jurors [JUROR] --amount [AMOUNT] --from [FROM] --network [NETWORK] 

Reviewer 2: I get issues here regardless of whether I enter 9000 + 18 zeroes or 9000 as the amount:

```
node ./bin/index.js activate --juror 0xd977dA63d086d222EDE0aa68ee84328310485FFE --amount 9000000000000000000000 --from 0xd977dA63d086d222EDE0aa68ee84328310485FFE --network usability
🚨  Process finished with error:
{ Error: transaction failed (transactionHash="0xcc02e19c4cbd95cd999589a0b1964e2b70542b7553a02a834a2bb3c6abdb97f9", transaction={"hash":"0xcc02e19c4cbd95cd999589a0b1964e2b70542b7553a02a834a2bb3c6abdb97f9","blockHash":null,"blockNumber":null,"transactionIndex":null,"confirmations":0,"from":"0xd977dA63d086d222EDE0aa68ee84328310485FFE","gasPrice":{"_hex":"0x37e11d601"},"gasLimit":{"_hex":"0x694920"},"to":"0x8FFD0Adb0Fc03a270dA84b6d10F87ec251296A50","value":{"_hex":"0x0"},"nonce":625,"data":"0xb260c42a0000000000000000000000000000001a72d8f32fd36112e75a900a8000000000","r":"0xadd30ffa70a84c4768ea1b9def662df279045eef36ed2d0e8f127b15d9f6170a","s":"0x78e370ee2b9679db447c7162e298ac53db249bcd7f69d61305c9ab66a7ea2a60","v":28,"creates":null,"raw":"0xf88b82027185037e11d60183694920948ffd0adb0fc03a270da84b6d10f87ec251296a5080a4b260c42a0000000000000000000000000000001a72d8f32fd36112e75a900a80000000001ca0add30ffa70a84c4768ea1b9def662df279045eef36ed2d0e8f127b15d9f6170aa078e370ee2b9679db447c7162e298ac53db249bcd7f69d61305c9ab66a7ea2a60","networkId":0,"chainId":0}, version=4.0.47)
    at Object.throwError (/home/light/court-backend/packages/shared/node_modules/ethers/errors.js:76:17)
    at /home/light/court-backend/packages/shared/node_modules/ethers/providers/base-provider.js:808:28
    at process._tickCallback (internal/process/next_tick.js:68:7)
  reason: 'transaction failed',
  code: 'CALL_EXCEPTION',
  transactionHash:
   '0xcc02e19c4cbd95cd999589a0b1964e2b70542b7553a02a834a2bb3c6abdb97f9',
  transaction:
   { hash:
      '0xcc02e19c4cbd95cd999589a0b1964e2b70542b7553a02a834a2bb3c6abdb97f9',
     blockHash: null,
     blockNumber: null,
     transactionIndex: null,
     confirmations: 0,
     from: '0xd977dA63d086d222EDE0aa68ee84328310485FFE',
     gasPrice: BigNumber { _hex: '0x37e11d601' },
     gasLimit: BigNumber { _hex: '0x694920' },
     to: '0x8FFD0Adb0Fc03a270dA84b6d10F87ec251296A50',
     value: BigNumber { _hex: '0x0' },
     nonce: 625,
     data:
      '0xb260c42a0000000000000000000000000000001a72d8f32fd36112e75a900a8000000000',
     r:
      '0xadd30ffa70a84c4768ea1b9def662df279045eef36ed2d0e8f127b15d9f6170a',
     s:
      '0x78e370ee2b9679db447c7162e298ac53db249bcd7f69d61305c9ab66a7ea2a60',
     v: 28,
     creates: null,
     raw:
      '0xf88b82027185037e11d60183694920948ffd0adb0fc03a270da84b6d10f87ec251296a5080a4b260c42a0000000000000000000000000000001a72d8f32fd36112e75a900a80000000001ca0add30ffa70a84c4768ea1b9def662df279045eef36ed2d0e8f127b15d9f6170aa078e370ee2b9679db447c7162e298ac53db249bcd7f69d61305c9ab66a7ea2a60',
     networkId: 0,
     chainId: 0,
     wait: [Function] } }
```
or

```
node ./bin/index.js activate --juror 0xd977dA63d086d222EDE0aa68ee84328310485FFE --amount 9000 --from 0xd977dA63d086d222EDE0aa68ee84328310485FFE --network usability
🚨  Process finished with error:
{ Error: transaction failed (transactionHash="0xad76ffe8aca32f15380af12e5a3102ab2d33ee4edba25b0f885649e4b39fce10", transaction={"hash":"0xad76ffe8aca32f15380af12e5a3102ab2d33ee4edba25b0f885649e4b39fce10","blockHash":null,"blockNumber":null,"transactionIndex":null,"confirmations":0,"from":"0xd977dA63d086d222EDE0aa68ee84328310485FFE","gasPrice":{"_hex":"0x37e11d601"},"gasLimit":{"_hex":"0x694920"},"to":"0x8FFD0Adb0Fc03a270dA84b6d10F87ec251296A50","value":{"_hex":"0x0"},"nonce":626,"data":"0xb260c42a0000000000000000000000000000000000000000000001e7e4171bf4d3a00000","r":"0xd762735d9a379d5db3fd10d6dbcdd37c64588606f41c60de62c0c6572e58ad78","s":"0x6d7a26a001e4703ef961030f2af374aeae20305a8272404a278898d3f57f6a45","v":28,"creates":null,"raw":"0xf88b82027285037e11d60183694920948ffd0adb0fc03a270da84b6d10f87ec251296a5080a4b260c42a0000000000000000000000000000000000000000000001e7e4171bf4d3a000001ca0d762735d9a379d5db3fd10d6dbcdd37c64588606f41c60de62c0c6572e58ad78a06d7a26a001e4703ef961030f2af374aeae20305a8272404a278898d3f57f6a45","networkId":0,"chainId":0}, version=4.0.47)
    at Object.throwError (/home/light/court-backend/packages/shared/node_modules/ethers/errors.js:76:17)
    at /home/light/court-backend/packages/shared/node_modules/ethers/providers/base-provider.js:808:28
    at process._tickCallback (internal/process/next_tick.js:68:7)
  reason: 'transaction failed',
  code: 'CALL_EXCEPTION',
  transactionHash:
   '0xad76ffe8aca32f15380af12e5a3102ab2d33ee4edba25b0f885649e4b39fce10',
  transaction:
   { hash:
      '0xad76ffe8aca32f15380af12e5a3102ab2d33ee4edba25b0f885649e4b39fce10',
     blockHash: null,
     blockNumber: null,
     transactionIndex: null,
     confirmations: 0,
     from: '0xd977dA63d086d222EDE0aa68ee84328310485FFE',
     gasPrice: BigNumber { _hex: '0x37e11d601' },
     gasLimit: BigNumber { _hex: '0x694920' },
     to: '0x8FFD0Adb0Fc03a270dA84b6d10F87ec251296A50',
     value: BigNumber { _hex: '0x0' },
     nonce: 626,
     data:
      '0xb260c42a0000000000000000000000000000000000000000000001e7e4171bf4d3a00000',
     r:
      '0xd762735d9a379d5db3fd10d6dbcdd37c64588606f41c60de62c0c6572e58ad78',
     s:
      '0x6d7a26a001e4703ef961030f2af374aeae20305a8272404a278898d3f57f6a45',
     v: 28,
     creates: null,
     raw:
      '0xf88b82027285037e11d60183694920948ffd0adb0fc03a270da84b6d10f87ec251296a5080a4b260c42a0000000000000000000000000000000000000000000001e7e4171bf4d3a000001ca0d762735d9a379d5db3fd10d6dbcdd37c64588606f41c60de62c0c6572e58ad78a06d7a26a001e4703ef961030f2af374aeae20305a8272404a278898d3f57f6a45',
     networkId: 0,
     chainId: 0,
     wait: [Function] } }
```

Author: It is without the decimals, added more details to make that clear now

```

Where:
- `[JUROR]`: address of the juror you will activate the tokens for
- `[AMOUNT]`: amount of fake ANJ tokens you will activate for the specified juror
- `[FROM]`: address paying for the fake ANJ tokens; this must be the address you used to claim the tokens from the faucet
- `[NETWORK]`: name of the Aragon Court instance you are willing to use: `usability`, `rinkeby`, or `ropsten` 

In case the sender address is the juror to be activated, you can simplify to the following:
```bash

> node ./bin/index.js activate -a [AMOUNT] -n [NETWORK]

Reviewer 2: simplified command results in the following error:

```
node ./bin/index.js activate -a 9000 -n usability
🚨  Process finished with error:
Error: insufficient funds for gas * price + value
    at /home/light/court-backend/packages/shared/node_modules/web3-provider-engine/subproviders/provider.js:18:36
    at XMLHttpRequest.request.onreadystatechange (/home/light/court-backend/packages/shared/node_modules/web3-providers-http/src/index.js:96:13)
    at XMLHttpRequestEventTarget.dispatchEvent (/home/light/court-backend/packages/shared/node_modules/xhr2-cookies/xml-http-request-event-target.ts:44:13)
    at XMLHttpRequest._setReadyState (/home/light/court-backend/packages/shared/node_modules/xhr2-cookies/xml-http-request.ts:219:8)
    at XMLHttpRequest._onHttpResponseEnd (/home/light/court-backend/packages/shared/node_modules/xhr2-cookies/xml-http-request.ts:345:8)
    at IncomingMessage.<anonymous> (/home/light/court-backend/packages/shared/node_modules/xhr2-cookies/xml-http-request.ts:311:39)
    at IncomingMessage.emit (events.js:194:15)
    at endReadableNT (_stream_readable.js:1125:12)
    at process._tickCallback (internal/process/next_tick.js:63:19)
```

Author: This is probably due to how your key file is being parsed, I removed this step to avoid misleading errors

To continue with the testing guide you will need to use it. First, make sure you clone it and install its dependencies as follows:
```
git clone https://github.com/aragonone/court-backend/
cd court-backend
git checkout master
npm i
npx lerna bootstrap
cd packages/cli
```

This CLI tool is built on top of Truffle using a custom [config file](https://www.npmjs.com/package/@aragon/truffle-config-v5) provided by Aragon.
Please review that package's documentation to understand how to set up your private keys for testing.

Let's continue with the Aragon Court testing guide and see how we can use the CLI tool.

## 8.4. Becoming a juror

To become a juror you simply need to activate some ANJ tokens into Aragon Court.
First make sure to have claimed some fake ANJ tokens from the faucet corresponding to the Aragon Court instance you're willing to try.
For now, the testing instances require a minimum of 10,000 ANJ so make sure to have at least that amount. 
Then, you can activate tokens into Aragon Court using the `stake` and `activate` commands of the CLI tool as follows:

```bash
node ./bin/index.js stake --juror [JUROR] --amount [AMOUNT] --from [FROM] --network [NETWORK] --verbose

> node ./bin/index.js activate --juror [JUROR] --amount [AMOUNT] --from [FROM] --network [NETWORK] --verbose

Reviewer 2: The `stake` command worked but I ran into an error on the `activate` command:

```
node ./bin/index.js activate --juror 0xd977dA63d086d222EDE0aa68ee84328310485FFE --amount 9000 --from 0xd977dA63d086d222EDE0aa68ee84328310485FFE --network usability
🚨  Process finished with error:
{ Error: transaction failed (transactionHash="0x209334fdcb7846bcb86f7c95a22a024394300d774aac42819f9997cadb14e2ce", transaction={"hash":"0x209334fdcb7846bcb86f7c95a22a024394300d774aac42819f9997cadb14e2ce","blockHash":null,"blockNumber":null,"transactionIndex":null,"confirmations":0,"from":"0xd977dA63d086d222EDE0aa68ee84328310485FFE","gasPrice":{"_hex":"0x37e11d601"},"gasLimit":{"_hex":"0x694920"},"to":"0x8FFD0Adb0Fc03a270dA84b6d10F87ec251296A50","value":{"_hex":"0x0"},"nonce":637,"data":"0xb260c42a0000000000000000000000000000000000000000000001e7e4171bf4d3a00000","r":"0xc64e7ded1f4c4ddb65b522193123f48f52082db9f49abc727b29b201ff533d63","s":"0x2983c9f135a6599c34f2db6574449c4e142e599f0cd0ff8446c61cfa11577b0c","v":27,"creates":null,"raw":"0xf88b82027d85037e11d60183694920948ffd0adb0fc03a270da84b6d10f87ec251296a5080a4b260c42a0000000000000000000000000000000000000000000001e7e4171bf4d3a000001ba0c64e7ded1f4c4ddb65b522193123f48f52082db9f49abc727b29b201ff533d63a02983c9f135a6599c34f2db6574449c4e142e599f0cd0ff8446c61cfa11577b0c","networkId":0,"chainId":0}, version=4.0.47)
    at Object.throwError (/home/light/court-backend/packages/shared/node_modules/ethers/errors.js:76:17)
    at /home/light/court-backend/packages/shared/node_modules/ethers/providers/base-provider.js:808:28
    at process._tickCallback (internal/process/next_tick.js:68:7)
  reason: 'transaction failed',
  code: 'CALL_EXCEPTION',
  transactionHash:
   '0x209334fdcb7846bcb86f7c95a22a024394300d774aac42819f9997cadb14e2ce',
  transaction:
   { hash:
      '0x209334fdcb7846bcb86f7c95a22a024394300d774aac42819f9997cadb14e2ce',
     blockHash: null,
     blockNumber: null,
     transactionIndex: null,
     confirmations: 0,
     from: '0xd977dA63d086d222EDE0aa68ee84328310485FFE',
     gasPrice: BigNumber { _hex: '0x37e11d601' },
     gasLimit: BigNumber { _hex: '0x694920' },
     to: '0x8FFD0Adb0Fc03a270dA84b6d10F87ec251296A50',
     value: BigNumber { _hex: '0x0' },
     nonce: 637,
     data:
      '0xb260c42a0000000000000000000000000000000000000000000001e7e4171bf4d3a00000',
     r:
      '0xc64e7ded1f4c4ddb65b522193123f48f52082db9f49abc727b29b201ff533d63',
     s:
      '0x2983c9f135a6599c34f2db6574449c4e142e599f0cd0ff8446c61cfa11577b0c',
     v: 27,
     creates: null,
     raw:
      '0xf88b82027d85037e11d60183694920948ffd0adb0fc03a270da84b6d10f87ec251296a5080a4b260c42a0000000000000000000000000000000000000000000001e7e4171bf4d3a000001ba0c64e7ded1f4c4ddb65b522193123f48f52082db9f49abc727b29b201ff533d63a02983c9f135a6599c34f2db6574449c4e142e599f0cd0ff8446c61cfa11577b0c',
     networkId: 0,
     chainId: 0,
     wait: [Function] } }
```

Reviewer 2: Nevermind just saw this above:

> For now, the testing instances require a minimum of 10,000 ANJ so make sure to have at least that amount.

Note that you can also avoid the flag `--verbose` if you want to avoid having too much details about the transactions being sent to the network.

You can check your current stake as a juror in the dashboards linked above in section 8.1.

## 8.5. Creating a dispute

As you may know, disputes can only be submitted to Aragon Court through smart contracts that implement a specific interface to support being ruled by the court itself.
This is specified by the [`IArbitrable`](../../contracts/arbitration/IArbitrable.sol) interface.

Thus, the first thing we should do is to deploy an Arbitrable contract. You can do this from the CLI running the following command:

```bash
node ./bin/index.js arbitrable -n [NETWORK] --verbose
```

This command will output the address of your new Arbitrable contract.

The next step is to subscribe your Arbitrable instance to Aragon Court by paying the subscription fees.
For the testing instances, each subscription period costs 7,500 fake-DAI, so make sure you claim some fake DAI from the ERC20 faucet as described in section 8.2.

Once you have done that you can subscribe your Arbitrable instance running the following command:

```bash

> node ./bin/index.js subscribe -a [ARBITRABLE] -n [NETWORK] --verbose

Reviewer 2: ran into an error on this step. I have 10,000 usability dai in my address to pay for the subscription, and the address in the command is the output of the previous command to generate an arbitrable contract

```
node ./bin/index.js subscribe -a 0x3233fCA0bFd0ED767594b2dF933fd8f253B9D756 -n usability
🚨  Process finished with error:
{ Error: transaction failed (transactionHash="0xd492d4c1cf3a399b2ac0077039373e89adfecd7901aa425d6629c78091f50023", transaction={"hash":"0xd492d4c1cf3a399b2ac0077039373e89adfecd7901aa425d6629c78091f50023","blockHash":null,"blockNumber":null,"transactionIndex":null,"confirmations":0,"from":"0x9DD224a54849bbd278595903300C9552c4a3E841","gasPrice":{"_hex":"0x37e11d601"},"gasLimit":{"_hex":"0x694920"},"to":"0x7350083Bac0e46F9CE1872Bfb4d7D58A42BA7216","value":{"_hex":"0x0"},"nonce":35,"data":"0xde974ba40000000000000000000000003233fca0bfd0ed767594b2df933fd8f253b9d7560000000000000000000000000000000000000000000000000000000000000001","r":"0x1c78cdf57fd85b311ceed94dba9fdacbaad1255b883d4b10d2232125616e3516","s":"0x5f2e90be3148a56c22f0341d021d21f6507ee1fcb45cbadcc9721be85b1d6d8a","v":28,"creates":null,"raw":"0xf8aa2385037e11d60183694920947350083bac0e46f9ce1872bfb4d7d58a42ba721680b844de974ba40000000000000000000000003233fca0bfd0ed767594b2df933fd8f253b9d75600000000000000000000000000000000000000000000000000000000000000011ca01c78cdf57fd85b311ceed94dba9fdacbaad1255b883d4b10d2232125616e3516a05f2e90be3148a56c22f0341d021d21f6507ee1fcb45cbadcc9721be85b1d6d8a","networkId":0,"chainId":0}, version=4.0.47)
    at Object.throwError (/home/light/court-backend/packages/shared/node_modules/ethers/errors.js:76:17)
    at /home/light/court-backend/packages/shared/node_modules/ethers/providers/base-provider.js:808:28
    at process._tickCallback (internal/process/next_tick.js:68:7)
  reason: 'transaction failed',
  code: 'CALL_EXCEPTION',
  transactionHash:
   '0xd492d4c1cf3a399b2ac0077039373e89adfecd7901aa425d6629c78091f50023',
  transaction:
   { hash:
      '0xd492d4c1cf3a399b2ac0077039373e89adfecd7901aa425d6629c78091f50023',
     blockHash: null,
     blockNumber: null,
     transactionIndex: null,
     confirmations: 0,
     from: '0x9DD224a54849bbd278595903300C9552c4a3E841',
     gasPrice: BigNumber { _hex: '0x37e11d601' },
     gasLimit: BigNumber { _hex: '0x694920' },
     to: '0x7350083Bac0e46F9CE1872Bfb4d7D58A42BA7216',
     value: BigNumber { _hex: '0x0' },
     nonce: 35,
     data:
      '0xde974ba40000000000000000000000003233fca0bfd0ed767594b2df933fd8f253b9d7560000000000000000000000000000000000000000000000000000000000000001',
     r:
      '0x1c78cdf57fd85b311ceed94dba9fdacbaad1255b883d4b10d2232125616e3516',
     s:
      '0x5f2e90be3148a56c22f0341d021d21f6507ee1fcb45cbadcc9721be85b1d6d8a',
     v: 28,
     creates: null,
     raw:
      '0xf8aa2385037e11d60183694920947350083bac0e46f9ce1872bfb4d7d58a42ba721680b844de974ba40000000000000000000000003233fca0bfd0ed767594b2df933fd8f253b9d75600000000000000000000000000000000000000000000000000000000000000011ca01c78cdf57fd85b311ceed94dba9fdacbaad1255b883d4b10d2232125616e3516a05f2e90be3148a56c22f0341d021d21f6507ee1fcb45cbadcc9721be85b1d6d8a',
     networkId: 0,
     chainId: 0,
     wait: [Function] } }
```

Author: @john-light added the sender in all these steps to avoid having issues with how the key store is being loaded

- `[SUBMITTER_N]`: addresses submitting each piece of evidence; this list should match the evidence list length 
- `-c` flag: optional to declare that the evidence submission period should be immediately closed. Otherwise you will need to manually close it afterwards. 
- `[FROM]`: address owning the Arbitrable instance being called; this address must be the one you used to delay the Arbitrable instance before
- `[NETWORK]`: name of the Aragon Court instance you are using: `usability`, `rinkeby`, or `ropsten`

This command will output the ID of the dispute you've just created.

A few things to bear in mind is that, even though the `[METADATA]` and `[EVIDENCE_N]` arguments could be any arbitrary information, in order to use the Court Dashboard to rule disputes, these should follow a specific structure.
Please check out the [Court Dashboard instructions](https://github.com/aragon/court-dashboard/blob/development/docs/metadata.md) about how these objects should be formatted.

### 8.5.1. Metadata

The Court Dashboard expects the metadata to be built as a specific JSON formatted data:
```json
{ 
    "description": "Your dispute description",  
    "metadata": "[CID]/metadata.json" 

> }

Reviewer 2: is this the content of the metadata.json file? (I cannot tell because the example you link to below does not work, as my other comment says)

or do I create this jason blob _after_ creating the metadata.json file, and paste this into the `[METADATA]` argument of the `dispute` command?

even after reading https://github.com/aragon/court-dashboard/blob/development/docs/metadata.md I am pretty lost already

Author: This is what you should paste into `[METADATA]`, the link shows an example of what you should upload to IPFS which will basically tell the `CID` of the metadata json file

```

Where `[CID]` is the hash of a directory holding a `metadata.json` file with the rest of the metadata.

> You can checkout the [`metadata.json`](https://ipfs.io/ipfs/QmYt33BkuHMLe4dSTfLan7QXxQVTGRyhVLt5sujrZkhd1w/metadata.json) file used for the [dispute #1](https://court-usability.aragon.org/disputes/1) of Aragon Court as an example.

Reviewer 2: this metadata.json file link is not loading for me

Author: Maybe we can paste the metadata here, cause IPFS links are a bit flaky

# 8. Testing guide

This guide aims to cover all the things you should know in order to try Aragon Court or integrate your application with it. 

## 8.1. Testing instances

There are a few testing instances already deployed for Aragon Court. 
All of these are mimicking the mainnet instance with some exceptions of term durations to provide a better testing experience.
Additionally, all the instances are using their own deployed version of the following ERC20 tokens:
- ANJ, the native token of Aragon Court. You will need some fake ANJ to stake as a juror to be selected to resolve disputes.  
- DAI, used for the Aragon Court fees. You will need some fake DAI to pay jurors and pay the Court subscription fees.

Of course, there is an ERC20 faucet deployed for all these instances that you can use to claim some fake ANJ or DAI to start testing. More information is outlined below on using these faucets.

### 8.1.1. Usability

This is probably the most useful testing instance you would like to try. 
Fees are low and court terms last a few minutes to make sure you can interact with it a bit faster.

- Network: Rinkeby

> - Court term: 30 minutes

Reviewer 2: Feedback from a UX perspective: 30 minute terms are still a really long time. Commit vote is a 3 hour period, Reveal vote is another 3 hours, appeal is another 3 hours, etc etc... it's taking me all day to test and I have to set manual timers to remind me when to go back and do something since there's no email notifications. I think we should drop the term period down to like literally a few minutes (no more than 2-3 minutes long) for the usability instance so that users can create a dispute and run through the whole thing in less than an hour.

Author: For the usability instance the commit period last 24 terms, we can reduce that to 1 term to make it faster

Reviewer 2: For me, reducing all periods to one term would be ideal. Then I can finish testing a whole case even with maximum number of appeals in a few hours instead of days.

- Subscription period: 90 court terms (1 month)
- Dashboard: https://court-ropsten.aragon.org/
- Address: [`0x3b26bc496aebaed5b3e0e81cde6b582cde71396e`](http://ropsten.etherscan.io/address/0x3b26bc496aebaed5b3e0e81cde6b582cde71396e)
- Fake ANJ: [`0xc863e1ccc047beff17022f4229dbe6321a6bce65`](http://ropsten.etherscan.io/address/0xc863e1ccc047beff17022f4229dbe6321a6bce65)
- Fake DAI: [`0x4e1f48db14d7e1ada090c42ffe15ff3024eec8bf`](http://ropsten.etherscan.io/address/0x4e1f48db14d7e1ada090c42ffe15ff3024eec8bf)
- ERC20 faucet: [`0x83c1ECDC6fAAb783d9e3ac2C714C0eEce3349638`](http://ropsten.etherscan.io/address/0x83c1ECDC6fAAb783d9e3ac2C714C0eEce3349638)

### 8.1.4. Local

To deploy a local instance of Aragon Court you will need to clone the deployment scripts first:
 
```bash
git clone https://github.com/aragon/aragon-network-deploy/
cd aragon-network-deploy
npm i 
```

Once you have done that, make sure you have a local Ganache running:

```bash
npx ganache-cli -i 15 --port 8545 --gasLimit 8000000 --deterministic
```

Then, open a separate terminal in the same directory of the scripts repo and deploy a local instance by running the following command:

```bash

> npm run deploy:court:rpc

Reviewer 2: I get some errors on this step:

```
npm run deploy:court:rpc

> @aragon/court-deploy@1.2.0 deploy:court:rpc /home/light/aragon-network-deploy
> rm -f data/output/{minime,court}.rpc.json && npm run deploy:dai -- -n rpc && npm run deploy:anj -- -n rpc && npm run deploy:court -- -n rpc

> @aragon/court-deploy@1.2.0 deploy:dai /home/light/aragon-network-deploy
> node ./bin/index.js deploy-minime -i ./test/data/input/dai.js --verbose "-n" "rpc"

[BaseDeployer]                  ⚠️   Using previous deploy at /home/light/aragon-network-deploy/data/output/minime.rpc.json
[MiniMeDeployer]                ⚠️   Using previous deployed DAI instance at 0x5b1869D9A4C187F2EAa108f3062412ecf0526b24
[Error Handler]                 🚨  Process finished with error:
Error: Cannot create instance of MiniMeToken; no code at address 0x5b1869D9A4C187F2EAa108f3062412ecf0526b24
    at Object.checkCode (/home/light/aragon-network-deploy/node_modules/@truffle/contract/lib/utils.js:277:13)
    at Function.at (/home/light/aragon-network-deploy/node_modules/@truffle/contract/lib/contract/constructorMethods.js:50:19)
    at process._tickCallback (internal/process/next_tick.js:68:7)

npm ERR! code ELIFECYCLE
npm ERR! errno 1
npm ERR! @aragon/court-deploy@1.2.0 deploy:dai: `node ./bin/index.js deploy-minime -i ./test/data/input/dai.js --verbose "-n" "rpc"`
npm ERR! Exit status 1
npm ERR! 
npm ERR! Failed at the @aragon/court-deploy@1.2.0 deploy:dai script.
npm ERR! This is probably not a problem with npm. There is likely additional logging output above.

npm ERR! A complete log of this run can be found in:
npm ERR!     /home/light/.npm/_logs/2020-05-19T23_23_44_940Z-debug.log
npm ERR! code ELIFECYCLE
npm ERR! errno 1
npm ERR! @aragon/court-deploy@1.2.0 deploy:court:rpc: `rm -f data/output/{minime,court}.rpc.json && npm run deploy:dai -- -n rpc && npm run deploy:anj -- -n rpc && npm run deploy:court -- -n rpc`
npm ERR! Exit status 1
npm ERR! 
npm ERR! Failed at the @aragon/court-deploy@1.2.0 deploy:court:rpc script.
npm ERR! This is probably not a problem with npm. There is likely additional logging output above.
```

Author: Fixed now! Please make sure to pull the changes before trying again :)

```bash
node ./bin/index.js subscribe -a [ARBITRABLE] -f [FROM] -n [NETWORK] --verbose
``` 

Where:
- `[ARBITRABLE]`: address of the Arbitrable instance you deployed in the previous step
- `[FROM]`: address paying for the fake DAI tokens; this must be the address you used to claim the tokens from the faucet
- `[NETWORK]`: name of the Aragon Court instance you are using: `usability`, `rinkeby`, or `ropsten` 

Now, we are almost ready to create a dispute. The last step is to send some fake DAI to the Arbitrable instance so that it can pay for the court's dispute fees.
These are different from the subscription fees. The dispute fees are to pay the jurors for each dispute to be resolved.
For the testing instances, each dispute costs 30.87 fake-DAI (`30870000000000000000` with 18 decimals).
Thus, you will need to make a transfer from your account to your Arbitrable instance.
To do that you can use the Etherscan interface for the fake DAI instance linked in section 8.1.

Finally, we are ready to create your dispute running the following command:

```bash
node ./bin/index.js dispute \
  -a [ARBITRABLE] \
  -m [METADATA] \

>   -e [EVIDENCE_1] [EVIDENCE_2] ... [EVIDENCE_N] \

Reviewer 2: Worth noting here that the term "evidence" has been changed to "argument".

  -s [SUBMITTER_1] [SUBMITTER_1] ... [SUBMITTER_N] \
  -c \
  -f [FROM] \
  -n [NETWORK] \
  --verbose
```

Where: 
- `[ARBITRABLE]`: address of your Arbitrable instance
- `[METADATA]`: metadata to be linked for your dispute (continue reading to have a better understanding of how to build a proper dispute metadata)
- `[EVIDENCE_N]`: reference to a human-readable evidence (continue reading to have a better understanding of how to provide a proper evidence reference) 
- `[SUBMITTER_N]`: addresses submitting each piece of evidence; this list should match the evidence list length 
- `-c` flag: optional to declare that the evidence submission period should be immediately closed. Otherwise, you will need to manually close it afterwards. 

> - `[FROM]`: address owning the Arbitrable instance being called; this address must be the one you used to delay the Arbitrable instance before

Reviewer 2: ```suggestion
- `[FROM]`: address owning the Arbitrable instance being called; this address must be the one you used to deploy the Arbitrable instance before
```

Author: too much time thinking about agreements 😂

}
```

Even though `agreementTitle`, `agreementText`, `plaintiff` and `defendant` are optional values, you will have a much better experience if you provide those.

Additionally, it is recommended to upload all these pieces of information together to IPFS. For example, you can take a look at [these files](./sample-dispute) we used to create this [sample dispute](https://court-usability.aragon.org/disputes/6). 
To upload those files we simply ran the following command while having the IPFS daemon running in background:

![ipfs](./images/ipfs-output.png)

As you can see, the `[METADATA_CID]` is the output marked in red, while the `[EVIDENCE_N_CID]`s are the ones in green.
Finally, following this example, this was the command we ran to create the dispute:

```bash
node ./bin/index.js dispute \
   -a 0x4BaBdb3aA57B351e57A9cb0e71B1c15Ec2e5479D \
   -m '{ "metadata": "QmbN3uaqRMWypUGKUbjuhL8wCgFXGgfktQgoKhTp6zUm6o/metadata.json" }' \
   -e ipfs:QmQn1eK9jbKQtwHoUwgXw3QP7dZe6rSDmyPut9PLXeHjhR ipfs:QmWRBN26uoL7MdZJWhSuBaKgCVvStBQMvFwSxtseTDY32S \
   -s 0x59d0b5475AcF30F24EcA10cb66BB5Ed75d3d9016 0x61F73dFc8561C322171c774E5BE0D9Ae21b2da42 \
   -c \
   -n usability \

>    --verbose

Reviewer 2: This is what I got when I tried to run this command:

```
node ./bin/index.js dispute \ -a 0x814f2bD1F175919186253054af55F7dBe5709C65 \ -m '{ "metadata": "QmQVpaKU5G6fu8sg1FRiQtvPDJHV9jBx3awX1ttLEbzaPf" }' \ -e ipfs:Qmc9qKp57cPCH7R1B3KE3QfEHxmw8NqCmHZNKtHcY2noSg ipfs:QmeioSQZdmtzTtPpC4tffZ7ziW7CkSifEyEB6pikryGZt1 \ -s 0xd977dA63d086d222EDE0aa68ee84328310485FFE 0x9DD224a54849bbd278595903300C9552c4a3E841 \ -c \ -f 0xd977dA63d086d222EDE0aa68ee84328310485FFE \ -n usability \ --verbose
index.js dispute

Create dispute submitting evidence

Options:
  --version         Show version number                                [boolean]
  --help            Show help                                          [boolean]
  --arbitrable, -a  Address of the arbitrable instance creating the dispute
                                                             [string] [required]
  --rulings, -r     Number of rulings         [string] [required] [default: "2"]
  --metadata, -m    Dispute metadata, it will only be logged            [string]
  --evidence, -e    Evidence contents                                    [array]
  --submitters, -s  Submitters for each evidence content                 [array]
  --close, -c       Whether the evidence period should be closed or not
                                                      [boolean] [default: false]
  --from, -f        Sender address                                      [string]
  --network, -n     Network name                             [string] [required]
  --silent          No logging                        [boolean] [default: false]
  --verbose         Verbose logging mode              [boolean] [default: false]

Missing required arguments: arbitrable, network
```

Author: I saw you were able to submit a dispute, am I right?

Reviewer 2: @facuspagnuolo yes, the trick was to remove the backslashes. I recommend removing the backslashes from the docs.

Please check out the [Court Dashboard instructions](https://github.com/aragon/court-dashboard/blob/development/docs/metadata.md) about how these objects should be formatted.

### 8.5.1. Metadata

The Court Dashboard expects the metadata to be built as a specific JSON formatted data:
```json
{ 
    "description": "Your dispute description",  
    "metadata": "[CID]/metadata.json" 
}
```

Where `[CID]` is the hash of a directory holding a `metadata.json` file with the rest of the metadata.
You can checkout the [`metadata.json`](https://ipfs.io/ipfs/QmYt33BkuHMLe4dSTfLan7QXxQVTGRyhVLt5sujrZkhd1w/metadata.json) file used for the [dispute #1](https://court-usability.aragon.org/disputes/1) of Aragon Court as an example.

### 8.5.2. Evidence

For the Court Dashboard, the evidence is simpler than the metadata, it only needs to be a human-readable content uploaded to IPFS.
Then it should be submitted as `ipfs:[CID]`. For example, [`ipfs:QmYGNe8jhTEwdDfixtDnPpzjQpXhX2nMj3xMK3swy69naP`](https://ipfs.io/ipfs/QmYGNe8jhTEwdDfixtDnPpzjQpXhX2nMj3xMK3swy69naP) is the evidence submitted for the [dispute #1](https://court-usability.aragon.org/disputes/1) of Aragon Court.

## 8.6. Ruling a dispute

> You can use any of the Court Dashboard instances linked in section 8.1 to interact with your created disputes (note that in some environments, it may be difficult to ensure that your account is drafted—and therefore can be difficult to come to a ruling you want—due to the randomness nature of the court). 

Reviewer: ```suggestion
You can use any of the Court Dashboard instances linked in section 8.1 to interact with your created disputes (note that in some environments, it may be difficult to ensure that your account is drafted due to the randomness nature of the court—and therefore can be difficult to come to a ruling you want). 
```

> If you did not format or ensure your dispute metadata was available as explained in sections 8.5.1 and 8.5.2, the dispute will most likely not display the intended information to jurors.

Reviewer: ```suggestion
If your dispute's metadata was not correctly formatted or made available as explained in sections 8.5.1 and 8.5.2, the dispute will most likely not display the intended information to jurors.
```
